### ISSUE TITLE
Performance Issues

### ISSUE THREAD

**Issue by AugmentBLU** [swup/swup#120]
I have quite a complex web app that is ajax heavy and dynamic. I'm finding the after several changes to the dom or navigating pages, especially where there's tabs that there's a noticeable decrease in performance.  I have cache set to off and preload on and seems to slow after navigating around a few times. If I stay on a given page, without triggering Swup it is fine. Any idea what could be the issue? 

Also noticed that even if cache is set to off that after navigating to other pages that existing JS functions that shouldn't be available are still triggering - I realise that JS is probably in memory but shouldn't the cache set to off unload all JS or am I getting that wrong?

**Comment by timhere** [comment]
Hi @AugmentBLU, I have exactly the same issue occuring around a similar implementation. I'm using an event listener to reload scripts on page changes as recommended [here (#41)](https://github.com/gmrchk/swup/issues/41#issuecomment-422199791).

Navigating between several pages causes the whole site to become increasingly sluggish, with cpu pegging at 100% and fps dropping into single figures. Staying on the same page is fine.

Setting `cache: false` has no discernable effect.

**Comment by gmrchk** [comment]
Hi guys,

Swup cache has nothing to do with JS at all actually. The option only defines whether swup should store the HTML of already loaded pages or not. In case that cache is enabled, swup can make fewer requests to your server and in the repeated visits of some pages, the load is basically instant. 

If you are having trouble with performance, it means you are attaching too many listeners (or creating too many instances of whatever you have on the page), without removing previous. 

For each page that is visited with swup, the JavaScript for the previous should be manually disabled (swups `willReplaceContent` event can help with that). Fortunately, in case a listener is attached to an element that is removed from DOM by swup, all of it's attached JS should be removed from memory by browser automatically. However, some listeners, like the one for `scroll` event, are usually attached to the window object, which never gets replaced by swup. That means with every page transition, you're could be adding one extra very costly event listener. The same goes for instance of some libraries...

TL/DR

Remove listeners and destroy instances of plugins on `willReplaceContent` swup event, before other listeners and instances are created on `contentReplaced` event.

**Comment by timhere** [comment]
Hi @gmrchk, thanks for your helpful reply with good info.

I'm now trying a different approach based on your advice. 

Rather than instantiating a class by simply calling `new MyClass()`, I create an instance as a property on an object, e.g. `obj.instance = new MyClass()` inside the `contentReplaced` event. Then inside swup's `willReplaceContent` event, I set `obj.instance = null`. My hope was that the garbage collector would then find the instance had been dereferenced and do it's thing. 

Unfortunately, this doesn't seem to be solving the issue with gradual performance degradation.

I also noticed that an additional [mousewheel event listener](https://github.com/gmrchk/swup/blob/33367a1072f8b5ebc646d7e0ca338a811b53a41d/src/modules/scrollTo.js#L49) is added with each page load when `scroll: true` is set in swup's options. 

I will keep trying!

**Comment by ParallelUniv3rse** [comment]
Hi, same issue here, especially noticeable in firefox for some reason. Cpu usage goes up 100% (~50% to ~150%) only when an instance of swup (swupjs) is created. Upon navigation it gets increasingly worse in increments of 50-70%.
I'm destroying bound events before the page is replaced.

**Comment by ParallelUniv3rse** [comment]
More info on this issue: 
I've investigated and narrowed down the problem to [this line in getDataFromHtml()](https://github.com/gmrchk/swup/blob/master/src/modules/getDataFromHtml.js#L6).
As soon as the innerHTML is appended the CPU usage jumps by 100% in firefox. It seems the `fakeDom` variable does not get garbage collected when you just create it and don't append to DOM as is the expected usage. I'm not sure if it's a memory leak or a bug in the garbage collector itself. 
Adding `fakeDom.innerHTML = ''` before the return statement brings CPU usage down to normal levels. (I assume that the root div element still stays in memory though)

**Comment by gmrchk** [comment]
Hi guys,

good job on the investigating! I will have a look and find out what's the best way to force the browser to remove the reference... but I can't replicate the issues you guys are having (not even in Firefox).
I'm also not quite sure why would extra DOM structure in memory (even if large) cause continuous big CPU usage in a first place.

**Comment by gmrchk** [comment]
Even tho I'm unable to reproduce the problem, it's pretty obvious that this could cause one. 
I have implemented @ParallelUniv3rse suggestion with clearing the `fakeDom` variable in v2. 

Creation of additional mouse wheel event listeners mentioned by @timhere should also be fixed (and moved to scroll plugin).

Closing, but please let me know if this is still a problem.

**Comment by AngusDV** [comment]
hi
@gmrchk 
I check your plugin and I have a question, please before answer read this page
 https://www.corewebvitals.io/pagespeed/fix-avoid-excessive-dom-size-lighthouse
I am creating a project with swup and I am afraid of DOM size and I am read this page 
https://www.w3schools.com/jsref/prop_nodelist_length.asp
you can get DOM size with document.body.childNodes.length;
when I am click on each link I am see ChildNodes are increased
if I had childNode over 1500 I think to get this error excessive-DOM size
